File: app/impl/EmailComposer.java
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package impl;

import com.google.inject.ImplementedBy;
import java.util.Set;
import models.assessment.LanguageInspection;
import models.enrolment.ExamEnrolment;
import models.enrolment.ExaminationEvent;
import models.enrolment.Reservation;
import models.exam.Exam;
import models.facility.ExamMachine;
import models.iop.CollaborativeExam;
import models.user.User;

@ImplementedBy(value = EmailComposerImpl.class)
public interface EmailComposer {
    /**
     * Message sent to student when review is ready.
     */
    void composeInspectionReady(User student, User reviewer, Exam exam);

    /**
     * Message sent to other inspectors when review is ready.
     */
    void composeInspectionMessage(User inspector, User sender, Exam exam, String msg);
    void composeInspectionMessage(User inspector, User sender, CollaborativeExam ce, Exam exam, String msg);

    /**
     * Weekly summary report
     */
    void composeWeeklySummary(User teacher);

    /**
     * Message sent to student when reservation has been made.
     */
    void composeReservationNotification(User student, Reservation reservation, Exam exam, Boolean isReminder);

    /**
     * Message sent to student when examination event has been selected.
     */
    void composeExaminationEventNotification(User student, ExamEnrolment enrolment, Boolean isReminder);

    /**
     * Message sent to student when examination event has been cancelled.
     */
    void composeExaminationEventCancellationNotification(User user, Exam exam, ExaminationEvent event);
    void composeExaminationEventCancellationNotification(Set<User> users, Exam exam, ExaminationEvent event);

    /**
     * Message sent to newly added inspectors.
     */
    void composeExamReviewRequest(User toUser, User fromUser, Exam exam, String message);

    /**
     * Message sent to student when reservation has been cancelled.
     */
    void composeReservationCancellationNotification(
        User student,
        Reservation reservation,
        String message,
        Boolean isStudentUser,
        ExamEnrolment enrolment
    );

    /**
     * Message sent to student when externally made reservation has been cancelled by hosting admin.
     */
    void composeExternalReservationCancellationNotification(Reservation reservation, String message);

    /**
     * Message sent to student when reservation has been changed.
     */
    void composeReservationChangeNotification(
        User student,
        ExamMachine previous,
        ExamMachine current,
        ExamEnrolment enrolment
    );

    /**
     * Message sent to student when he/she has been enrolled to a private exam.
     */
    void composePrivateExamParticipantNotification(User student, User fromUser, Exam exam);

    /**
     * Message sent to teacher when student has finished a private exam.
     */
    void composePrivateExamEnded(User toUser, Exam exam);

    /**
     * Message sent to teacher when student did not show up for a private exam.
     */
    void composeNoShowMessage(User toUser, User student, Exam exam);

    /**
     * Message sent to student when he did not show up for exam.
     */
    void composeNoShowMessage(User student, String examName, String courseCode);

    /**
     * Message sent to teacher when language inspection is finished.
     */
    void composeLanguageInspectionFinishedMessage(User toUser, User inspector, LanguageInspection inspection);

    /**
     * Message sent to teacher when collaborative exam is created in the system.
     */
    void composeCollaborativeExamAnnouncement(Set<String> emails, User sender, Exam exam);
}

File: app/services/mail/EmailComposer.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package services.mail

import com.google.inject.ImplementedBy
import models.assessment.LanguageInspection
import models.enrolment.{ExamEnrolment, ExaminationEvent, Reservation}
import models.exam.Exam
import models.iop.CollaborativeExam
import models.user.User

import scala.concurrent.duration.Duration

@ImplementedBy(value = classOf[EmailComposerImpl])
trait EmailComposer {

  /** Message sent to a student when a review is ready.
    */
  def composeInspectionReady(student: User, reviewer: Option[User], exam: Exam): Unit

  /** Message sent to other inspectors when a review is ready.
    */
  def composeInspectionMessage(inspector: User, sender: User, exam: Exam, msg: String): Unit
  def composeInspectionMessage(
      inspector: User,
      sender: User,
      ce: CollaborativeExam,
      exam: Exam,
      msg: String
  ): Unit

  /** Weekly summary report
    */
  def composeWeeklySummary(teacher: User): Unit

  /** Message sent to a student when a reservation has been made.
    */
  def composeReservationNotification(
      student: User,
      reservation: Reservation,
      exam: Exam,
      isReminder: Boolean
  ): Unit

  /** Message sent to a student when an examination event has been selected.
    */
  def composeExaminationEventNotification(
      student: User,
      enrolment: ExamEnrolment,
      isReminder: Boolean
  ): Unit

  /** Message sent to a student when an examination event has been canceled.
    */
  def composeExaminationEventCancellationNotification(
      user: User,
      exam: Exam,
      event: ExaminationEvent
  ): Unit
  def composeExaminationEventCancellationNotification(
      users: Set[User],
      exam: Exam,
      event: ExaminationEvent
  ): Unit

  /** Message sent to newly added inspectors.
    */
  def composeExamReviewRequest(toUser: User, fromUser: User, exam: Exam, message: String): Unit

  /** Message sent to a student when a reservation has been canceled.
    */
  def composeReservationCancellationNotification(
      student: User,
      reservation: Reservation,
      message: Option[String],
      isStudentUser: Boolean,
      enrolment: ExamEnrolment
  ): Unit

  /** Message sent to a student when externally made reservation has been canceled by hosting admin.
    */
  def composeExternalReservationCancellationNotification(
      reservation: Reservation,
      message: Option[String]
  ): Unit

  /** Message sent to a student when a reservation has been changed.
    */
  def composeReservationChangeNotification(
      current: Reservation,
      previous: Reservation
  ): Unit

  /** Message sent to a student when she has been enrolled to a private exam.
    */
  def composePrivateExamParticipantNotification(student: User, fromUser: User, exam: Exam): Unit

  /** Message sent to a teacher when a student has finished a private exam.
    */
  def composePrivateExamEnded(toUser: User, exam: Exam): Unit

  /** Message sent to a teacher when a student did not show up for a private exam.
    */
  def composeNoShowMessage(toUser: User, student: User, exam: Exam): Unit

  /** Message sent to a student when he did not show up for exam.
    */
  def composeNoShowMessage(student: User, examName: String, courseCode: String): Unit

  /** Message sent to teacher when language inspection is finished.
    */
  def composeLanguageInspectionFinishedMessage(
      toUser: User,
      inspector: User,
      inspection: LanguageInspection
  ): Unit

  /** Message sent to teacher when a collaborative exam is created in the system.
    */
  def composeCollaborativeExamAnnouncement(emails: Set[String], sender: User, exam: Exam): Unit

  /** Schedule an email action to run after a delay. The action runs in the background and errors
    * are logged.
    */
  def scheduleEmail(delay: Duration)(action: => Unit): Unit
}
